staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/conversion_test.go

```diff
@@ -660,3 +661,71 @@ func TestJSONRoundTrip(t *testing.T) {
 		})
 	}
 }
+
+func TestMemoryEqual(t *testing.T) {
+	testcases := []struct {
+		a interface{}
+		b interface{}
+	}{
+		{apiextensions.JSONSchemaProps{}.XValidations, JSONSchemaProps{}.XValidations},
+	}
+
+	for _, tc := range testcases {
+		aType := reflect.TypeOf(tc.a)
+		bType := reflect.TypeOf(tc.b)
+		t.Run(aType.String(), func(t *testing.T) {
+			assertEqualTypes(t, nil, aType, bType)
+		})
+	}
+}
+
+func assertEqualTypes(t *testing.T, path []string, a, b reflect.Type) {
+	if a == b {
+		return
+	}
+
+	if a.Kind() != b.Kind() {
+		fatalTypeError(t, path, a, b, "mismatched Kind")
+	}
+
+	switch a.Kind() {
+	case reflect.Struct:
+		aFields := a.NumField()
+		bFields := b.NumField()
+		if aFields != bFields {
+			fatalTypeError(t, path, a, b, "mismatched field count")
+		}
+		for i := 0; i < aFields; i++ {
+			aField := a.Field(i)
+			bField := b.Field(i)
+			if aField.Name != bField.Name {
+				fatalTypeError(t, path, a, b, fmt.Sprintf("mismatched field name %d: %s %s", i, aField.Name, bField.Name))
+			}
+			if aField.Offset != bField.Offset {
+				fatalTypeError(t, path, a, b, fmt.Sprintf("mismatched field offset %d: %v %v", i, aField.Offset, bField.Offset))
+			}
+			if aField.Anonymous != bField.Anonymous {
+				fatalTypeError(t, path, a, b, fmt.Sprintf("mismatched field anonymous %d: %v %v", i, aField.Anonymous, bField.Anonymous))
+			}
+			if !reflect.DeepEqual(aField.Index, bField.Index) {
+				fatalTypeError(t, path, a, b, fmt.Sprintf("mismatched field index %d: %v %v", i, aField.Index, bField.Index))
+			}
+			path = append(path, aField.Name)
+			assertEqualTypes(t, path, aField.Type, bField.Type)
+			path = path[:len(path)-1]
+		}
+
+	case reflect.Ptr, reflect.Slice:
+		aElemType := a.Elem()
+		bElemType := b.Elem()
+		assertEqualTypes(t, path, aElemType, bElemType)
+
+	default:
+		fatalTypeError(t, path, a, b, "unhandled kind")
+	}
+}
+
+func fatalTypeError(t *testing.T, path []string, a, b reflect.Type, message string) {
+	t.Helper()
+	t.Fatalf("%s: %s: %s %s", strings.Join(path, ""), message, a, b)
```

Comment:
Nit ([same here](https://github.com/kubernetes/kubernetes/blob/3a132bd20694afce993efcc35b22b874e990dc4d/pkg/kubelet/cri/remote/conversion_test.go#L124)): 

```suggestion
	t.Fatalf("%s: %s: %s %s", strings.Join(path, "."), message, a, b)
```